cdap-app-fabric/src/test/java/co/cask/cdap/app/runtime/monitor/RuntimeMonitorTest.java

```diff
@@ -123,46 +143,71 @@ public void testRunTimeMonitor() throws Exception {
     cConfCopy.set(Constants.AppFabric.PROGRAM_STATUS_EVENT_TOPIC, "cdap-programStatus");
 
     RuntimeMonitor runtimeMonitor = new RuntimeMonitor(programRunId,
-                                                       cConfCopy, messagingService, clientConfigBuilder.build());
+                                                       cConfCopy, messagingService, clientConfigBuilder.build(),
+                                                       datasetFramework, transactionSystemClient);
+
+    runtimeMonitor.startAndWait();
+    // use different configuration for verification
+    String lastProcessed = verifyPublishedMessages(2, cConfCopy, 2, null);
+    runtimeMonitor.stopAndWait();
+
+    // publish some more messages to test offset manager
+    publishProgramStatus(programRunId);
+    verifyPublishedMessages(2, cConf, 2, lastProcessed);
 
+    runtimeMonitor = new RuntimeMonitor(programRunId, cConfCopy, messagingService, clientConfigBuilder.build(),
+                                        datasetFramework, transactionSystemClient);
     runtimeMonitor.startAndWait();
     // use different configuration for verification
-    verifyPublishedMessages(2, cConfCopy);
+    lastProcessed = verifyPublishedMessages(2, cConfCopy, 2, lastProcessed);
+
+    // publish completed status to trigger offset clean up
+    publishCompletedStatus(programRunId);
+
+    // use different configuration for verification
+    verifyPublishedMessages(2, cConfCopy, 1, lastProcessed);
 
     // wait for runtime server to stop automatically
     Tasks.waitFor(true, () -> !runtimeServer.isRunning(), 5, TimeUnit.MINUTES);
 
     runtimeMonitor.stopAndWait();
   }
 
-  private void verifyPublishedMessages(int limit, CConfiguration cConfig) throws Exception {
+  private String verifyPublishedMessages(int limit, CConfiguration cConfig, int expectedCount, final String messageId)
+    throws Exception {
     MessageFetcher fetcher = messagingContext.getMessageFetcher();
```

Comment:
You should now fetch transactionally to make sure the publishing side is doing the right thing.